warehouse/macaroons/security_policy.py

```diff
@@ -88,13 +88,22 @@ def identity(self, request):
         # Check to see if our Macaroon exists in the database, and if so
         # fetch the user that is associated with it.
         macaroon_service = request.find_service(IMacaroonService, context=None)
-        userid = macaroon_service.find_userid(macaroon)
 
-        if userid is None:
+        try:
+            dm = macaroon_service.find_from_raw(macaroon)
+        except InvalidMacaroonError:
             return None
 
-        user_service = request.find_service(IUserService, context=None)
-        return user_service.get_user(userid)
+        # Every Macaroon has either a user, or it's being used in the
+        # context of a project. For the latter, our authorization policy
+        # has already checked whether the project is valid for the Macaroon
+        # by checking the Macaroon's caveats.
+        if dm.user is not None:
+            return dm.user
+        else:
+            if not isinstance(request.context, Project):
+                return None
+            return request.context
```

Comment:
Flagging for review: in order to get this PR into a working state, we need to get the intended `Project` model as the request identity here. 


The current approach doesn't actually work, since the upload endpoint doesn't provide a `context`.

Some ideas:

* We could parse the request's body as a form here, and extract the project name from the metadata. This would essentially duplicate the work already being done in the actual upload endpoint, so it's not ideal.
* We could figure out how to set `request.context` for the upload route. This will be a little thorny, since project creation involves sending a project name that doesn't exist yet. That should never happen with a macaroon (since a user can't create a macaroon that references a nonexistent project), but it's more state to deal with.
* Something else?

Reply:
(And note: the reason we don't have `dm.project` to rely on is because a Macaroon can authenticate for multiple projects.)

Reply:
> That should never happen with a macaroon (since a user can't create a macaroon that references a nonexistent project), but it's more state to deal with.

I think I saw an issue recently where someone pointed out that you can have a Macaroon that refers to a now deleted project, unless we've fixed that and we're purging macaroons on deletion now.

Reply:
I take that back, it's a annoying to do:

* A macaroon might be scoped for multiple projects, meaning that we'd invalidate other potential users of the token
* The only server-side reference to the Macaroon's projects is kept in a JSON blob in the DB, so we'd need a query that peeks into that blob and checks whether the deleted project is a member of the projects list

Reply:
That being said, we definitely should purge or invalidate Macaroons on project (and account) deletion -- this PR makes users optional on Macaroons, giving us a potential security problem where an ephemeral orphaned macaroon could potentially authenticate for a project that's been deleted and then re-created by a different user. The window for that would be very small, but it is possible.

We could preempt that by ensuring that the project has an OIDC provider enabled, but that's a hack and not a real solution.

Reply:
https://github.com/pypi/warehouse/issues/9524 found the issue I was thinking of

Reply:
I haven't actually reviewed this PR yet, but are we allowing Macaroons that aren't attached to a user, and are attached to multiple projects?

Reply:
> I haven't actually reviewed this PR yet, but are we allowing Macaroons that aren't attached to a user, and are attached to multiple projects?

Yep, correct -- any project that authorizes the OIDC provider that verifies the JWT will be scoped within the Macaroon. That's based on the redesign described in https://github.com/pypi/warehouse/pull/11272#issuecomment-1168835516 and forwards.